### PR TITLE
Feat/yr/shopping cart

### PR DESCRIPTION
**Pull request by YgRamos9**
## Summary
- adding shopping cart

## Screenshot

https://user-images.githubusercontent.com/107204529/184993099-71fadeb1-65a8-4dc5-9e92-f3a1dacc311f.mov

## Trello Card
https://trello.com/c/kOhinBnm/12-carrito-de-compras

### CODE REVIEW

File: angular/src/app/components/book-card/book-card.component.html [hunk 1]
```diff
@@ -2,5 +2,12 @@
   <img class="book-cover" [src]="book.imageUrl" [alt]="book.title" />
   <span class="book-title text-ellipsis">{{ book.title }}</span>
   <span class="book-author text-ellipsis">{{ book.author }}</span>
-  <button class="cart-add" type="button">+</button>
+  <button
+    class="cart-add"
```
Comment (miguelhenaoa) [inline angular/src/app/components/book-card/book-card.component.html]:
```suggestion
    class="add-button"
```

File: angular/src/app/components/book-card/book-card.component.ts [hunk 2]
```diff
@@ -8,4 +9,10 @@ import { IBook } from 'src/app/interfaces/global.interface';
 })
 export class BookCardComponent {
   @Input() book!: IBook;
+
+  constructor(private shoppingService: ShoppingCartService) {}
+
+  addBook() {
```
Comment (miguelhenaoa) [inline angular/src/app/components/book-card/book-card.component.ts]: Add data type, in this case void
```suggestion
  addBook(): void {
```

File: angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.html [hunk 3]
```diff
@@ -0,0 +1,38 @@
+<button class="cart" data-testid="button-open-cart" (click)="showCart()">
+  <img src="assets/cart.png" />
```
Comment (miguelhenaoa) [inline angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.html]: alt property

File: angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.html [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  <span class="cart-count">{{ booksCount }}</span>
+</button>
+
+<div class="popup showPopup" *ngIf="openCart">
+  <div class="popup-content">
+    <img class="close" src="assets/close.png" (click)="showCart()" />
```
Comment (miguelhenaoa) [inline angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.html]: alt property

File: angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.html [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    <div class="popup-title">
+      Tu carrito de compras
+      <ng-container *ngIf="hasBooks; else elseBook">
+        <div class="popup-container" *ngFor="let book of books">
+          <div class="popup-content-details">
+            <div class="popup-content-item">
+              {{ book.title }} - {{ book.author }}
+              <img
```
Comment (miguelhenaoa) [inline angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.html:21]: same

File: angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.html [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+                src="assets/delete.png"
+              />
+            </div>
+            <div class="popup-content-item-line"></div>
+          </div>
+        </div>
+      </ng-container>
+      <ng-template #elseBook>
+        <div>
+          <img class="popup-empty-img" src="assets/cart-empty.png" />
```
Comment (miguelhenaoa) [inline angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.html]: same

File: angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.html [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+          <div class="popup-empty-title">El carrito está vacío</div>
+          <div class="popup-empty-desciption">
```
Comment (miguelhenaoa) [inline angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.html]: use span

File: angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.scss [hunk 8]
```diff
@@ -0,0 +1,125 @@
+@import 'src/styles/variables/colors';
+@import 'src/styles/variables/fonts';
+
+.cart {
+  cursor: pointer;
+  display: flex;
+  height: 24px;
+  margin: 0 29px;
+  position: relative;
+  width: 22.31px;
```
Comment (miguelhenaoa) [inline angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.scss]: round number

File: angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.scss [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  height: 56px;
+  justify-content: space-between;
+  line-height: 26px;
+  max-width: 100%;
+  width: 800px;
+}
+
+.popup-cover-delete {
+  height: 24px;
+  width: 19.49px;
```
Comment (miguelhenaoa) [inline angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.scss]: same

File: angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.scss [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+  background-color: $line-popup;
+  height: 2px;
+  max-width: 100%;
+  width: 800px;
+}
+
+.popup-empty-img {
+  height: 168px;
+  margin: 90px 233px 90px;
+  width: 230.66px;
```
Comment (miguelhenaoa) [inline angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.scss]: same

File: angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+export class ShoppingCartComponent implements OnInit, OnDestroy {
+  openCart: boolean = false;
+  books: IBook[] = [];
+  booksSubscription!: Subscription;
+  booksCount: number = 0;
+
+  constructor(public shoppingService: ShoppingCartService) {}
+
+  ngOnInit(): void {
+    this.booksSubscription = this.shoppingService
```
Comment (miguelhenaoa) [inline angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.ts]: shoppingSubscription?

File: angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        this.booksCount = response.length;
+      });
+  }
+
+  ngOnDestroy() {
+    this.booksSubscription.unsubscribe();
+  }
+
+  get hasBooks(): boolean {
+    return this.books.length > 0 ? true : false;
```
Comment (miguelhenaoa) [inline angular/src/app/screens/auth/components/shopping-cart/shopping-cart.component.ts]: Use booksCount

File: angular/src/app/services/shopping-cart.service.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+export class ShoppingCartService {
+  public booksSubject = new Subject<IBook[]>();
+  books: IBook[] = [];
+
+  setBook(book: IBook) {
+    this.books.push(book);
+    this.booksSubject.next(this.books);
+  }
+
+  getBooks() {
```
Comment (miguelhenaoa) [inline angular/src/app/services/shopping-cart.service.ts:17]: data type

**Review by miguelhenaoa — CHANGES_REQUESTED** [review]
Hi @YgRamos9 , here are some comments, check the data type of your functions, they are missing in many places

File: angular/src/app/services/shopping-cart.service.ts [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+@Injectable({
+  providedIn: 'root',
+})
+export class ShoppingCartService {
+  public booksSubject = new Subject<IBook[]>();
+  books: IBook[] = [];
+
+  setBook(book: IBook) {
+    this.books.push(book);
+    this.booksSubject.next(this.books);
```
Comment (joacomiguens) [inline angular/src/app/services/shopping-cart.service.ts]: create a new function that only do this thing 😄

File: angular/src/app/services/shopping-cart.service.ts [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+  }
+
+  getBooks() {
+    return this.booksSubject;
+  }
+
+  removeBook(id: number) {
+    this.books = this.books.filter((book) => book.id !== id);
```
Comment (miguelhenaoa) [inline angular/src/app/services/shopping-cart.service.ts:22]: Yoni, you don't need to create the books variable if you can iterate what you get from getBooks

**Review by miguelhenaoa — CHANGES_REQUESTED** [review]

File: angular/src/app/services/shopping-cart.service.ts (hunk 15 above)
Comment (YgRamos9) [reply angular/src/app/services/shopping-cart.service.ts:22]: Hello migue I was trying to carry out this process but for some unknown reason I cannot obtain the values ​​of said function I tried to carry out with a behaviorsubject and it did not work either.

**Review by joacomiguens — APPROVED** [review]
Great job bro 💯

**Review by miguelhenaoa — APPROVED** [review]

**Pull request merged**